src/libsyntax/parse/parser.rs

```diff
@@ -2921,6 +2937,27 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn is_at_start_of_range_notation_rhs(&self) -> bool {
+        if self.token.can_begin_expr() {
+            // parse `for i in 1.. { }` as infinite loop, not as `for i in (1..{})`.
+            if self.token == token::OpenDelim(token::Brace) {
+                return !self.restrictions.contains(RESTRICTION_NO_STRUCT_LITERAL);
+            }
+
+            // `1..*i` is ambiguous between `1..(*i)` and `(1..)*(i)`.
+            // We pick the `1..(*i)` interpretation.
+
+            // `r==1..&&true` is ambiguous between `r==(1..(&&true))` and `(r==(1..))&&true`.
+            // We pick the latter interpretation.
+            match self.token.to_binop() {
+                Some(op) => operator_prec(op) > range_prec,
+                None => true
+            }
+        } else {
+           false
+        }
+    }
+
```

Comment:
It looks like this ensures that `[a]..[b]` is still treated as `([a])..([b])` rather than `([a]..)[b]`, which is good. It's unlikely to break, but it might be worth putting a test in for that case anyway?